crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -784,10 +838,61 @@ where
                     }
 
                     self.visit_body(handler_body);
+                    // Each `except` block is mutually exclusive with all other `except` blocks.
+                    post_except_states.push(self.flow_snapshot());
+
+                    // It's unnecessary to do the `self.flow_restore()` call for the final except handler,
+                    // as we'll immediately call `self.flow_restore()` to a different state
+                    // as soon as this loop over the handlers terminates.
+                    if i < (num_handlers - 1) {
+                        self.flow_restore(pre_except_state.clone());
+                    }
                 }
 
+                // If we get to the `else` block, we know that 0 of the `except` blocks can have been executed,
+                // and the entire `try` block must have been executed:
+                self.flow_restore(post_try_block_state);
                 self.visit_body(orelse);
-                self.visit_body(finalbody);
+                let post_else_state = self.flow_snapshot();
+
+                // If there is a `finally` block, prepare to visit it:
+                let finally_snapshots = self.try_node_context_stack_mut().enter_finally_block();
+                debug_assert!(finally_snapshots
+                    .as_ref()
+                    .map_or(true, |_| !finalbody.is_empty()));
+
+                if let Some(finally_snapshots) = finally_snapshots {
+                    self.flow_restore(pre_try_block_state);
+                    for state in finally_snapshots {
+                        self.flow_merge(state);
+                    }
+
+                    // Visit the `finally` block!
+                    self.visit_body(finalbody);
+                }
+
+                // Prepare for exiting the `StmtTry` node
+                self.flow_restore(post_else_state);
+                for state in post_except_states {
+                    self.flow_merge(state);
+                }
+
+                // These definitions were erased by `self.flow_restore`ing to the post-`else` state.
+                // We can't simply `self.flow_merge()` with any snapshots taken during the `finally` block, however,
+                // as there are more potential definition states inside the `finally` block than there are
+                // from a point after the `finally` block's completion.
+                // Instead, we must manually re-add these definitions to the `use-def` map
+                if let Some(finally_definitions) = self.try_node_context_stack().pop_context() {
+                    for DefinitionRecord {
+                        symbol,
+                        definition,
+                        category,
+                    } in finally_definitions
+                    {
+                        self.current_use_def_map_mut()
+                            .record_definition(symbol, definition, category);
+                    }
+                }
```

Comment:
Ah, this is tricky indeed. I hadn't fully understood the awkward consequences of the way finally blocks work for our CFG. Thanks for taking the time to think this through!

Unfortunately I don't think this approach (of storing and then re-applying Definitions in the finally block) is going to give us the right results. Consider a case like this:

```
x = 1
try:
    x = could_raise_returns_str()
finally:
    y = x
reveal_type(y)
```

The correct revealed type for `y` is `str`, because in any case where code flow continues after the `finally`, that means the `try` block actually completed without an exception. But by storing and reapplying the `Definition` for `y`, we still get the understanding of the type of the RHS from the scenario where we might have an exception.

I think the only way to handle this correctly is to, in some form, duplicate or double visit the `finally` block. We effectively need to type it twice, once under the assumption that any code it protects might have raised, and again under the assumption that it didn't.

This will be a significant bit of work, as it troubles some core assumptions we have about visiting every expression exactly once. I don't think we should do it in this PR.

But I also don't think we should do this store-and-reapply-definitions thing, either. In the meantime, until/unless we get to a correct double-visit fix, I think the best tradeoff is to accept some false negatives while checking the `finally` block itself, but get the types correct after the `finally` block. In other words, for now I think we should just visit the finally block under the no-exceptions assumption.

What do you think?

Reply:
Thanks for the great example that shows the flaws in this approach! Ugh, I really thought I'd covered everything this time 🫠 This was, as I'm sure you guessed, the bit of this PR that I was least sure about.

> But I also don't think we should do this store-and-reapply-definitions thing, either, for two reasons. One is that I think it's just generally important for correctness that we maintain the control-flow-graph abstraction and don't work around it with tricks like this. The other is just about the tradeoff in semantics for Python code. Until/unless we get to a correct double-visit fix, I think the best tradeoff is to accept some false negatives while checking the `finally` block itself, but ensure we get the types correct after the `finally` block. In other words, for now I think we should just visit the finally block under the no-exceptions assumption.
> 
> What do you think?

I think this makes me a little sad after I spent so much time thinking about `finally` blocks 😆

I think it is pretty important that we fix this eventually. In the long run, this will lead to false positives as well as false negatives. For example, when we start emitting diagnostics for unreachable code, we will emit spurious errors on the `if` branch inside the `finally` block in this snippet, as we will incorrectly infer it as being unreachable:

```py
x = 42

try:
    x = could_raise_returns_int()
except:
    could_raise()
    x = "foo"
else:
    could_raise()
    x = "foo"
finally:
    if isinstance(x, int):
        ...  # we'd probably detect this as unreachable
             # unless we consider the fact that we might have jumped to the `finally`
             # branch from halfway through an `except` or `else` branch
    else:
        ...
```

Another way I thought of trying to fix this "awkwardness" was to utilise the fact that we know that `try`/`except` blocks with `finally` branches desugar to nested `try`/`except` blocks. We could attempt to "synthesize" a nested `StmtTry` node if we see that a `StmtTry` node has a non-empty `finally` suite. I actually started off trying to do that, but quickly stopped as this seemed like a simpler solution. (And I was also not sure how this would work with the assertions we have that you mentioned above, about only ever visiting every expression once.) Given the issue you just pointed out in your example, it seems like that probably is the only good way of doing it, though; there doesn't seem to be any way of taking shortcuts while respecting Python's semantics properly.

Reply:
Yeah, I think you're right that we will want to fix this.

Reply:
I've written up your edge case in my document describing control-flow semantics for exception handlers. It's _very_ specific! I believe it only applies to `StmtTry` nodes that:

- have `finally` blocks, and either:
  - do not have any `except` branches, or
  - all the `except` branches of the `StmtTry` node lead to immediate termination of the scope following the `finally` block, through either a `raise`, `return`or similar.

The specificity of the edge case doesn't mean that it's unimportant to consider, however.

Reply:
I think it also applies to try blocks with except handlers, it's just that the issue shifts to considering the possibility of an exception in the exception handler, rather than an exception in the try block?

And try/finally without except handlers is not an uncommon case.

Reply:
> I think it also applies to try blocks with except handlers, it's just that the issue shifts to considering the possibility of an exception in the exception handler, rather than an exception in the try block?

Ah, great point.

> And try/finally without except handlers is not an uncommon case.

I said _specific_, not uncommon! I agree that `try`/`finally` without `except` is pretty common, so I definitely agree this is an important case to consider.